model/ChangePermissionsService.php

```diff
@@ -0,0 +1,195 @@
+                $resource['permissions']['current'],
+                $permissionsDelta
+            );
+        }
+
+        return $this->deduplicateChanges($resources);
+    }
+
+    private function deduplicateChanges(array $resources): array
+    {
+        foreach ($resources as &$resource) {
+            foreach ($resource['permissions']['remove'] as &$removePermissions) {
+                $removePermissions = array_unique($removePermissions);
+            }
+
+            foreach ($resource['permissions']['add'] as &$addPermissions) {
+                $addPermissions = array_unique($addPermissions);
+            }
+        }
+
+        return $resources;
+    }
+
+    private function dryRun(array $resources): void
+    {
+        foreach ($resources as $resource) {
+            $newPermissions = $resource['permissions']['current'];
+
+            $this->dryRemove($resource, $newPermissions);
+            $this->dryAdd($resource, $newPermissions);
+
+            $this->assertHasUserWithGrantPermission($resource['id'], $newPermissions);
+        }
+    }
+
+    private function dryRemove(array $resource, array &$newPermissions): void
+    {
+        foreach ($resource['permissions']['remove'] as $user => $removePermissions) {
+            $newPermissions[$user] = array_diff(
+                $newPermissions[$user] ?? [],
+                $removePermissions
+            );
+        }
+    }
+
+    private function dryAdd(array $resource, array &$newPermissions): void
+    {
+        foreach ($resource['permissions']['add'] as $user => $addPermissions) {
+            $newPermissions[$user] = array_merge(
+                $resource['permissions']['current'][$user] ?? [],
+                $addPermissions
+            );
+        }
+    }
+
+    /**
+     * Checks if all resources after all actions are applied will have at least
+     * one user with GRANT permission.
+     */
+    private function assertHasUserWithGrantPermission(string $resourceId, array $newPermissions): void
+    {
+        foreach ($newPermissions as $permissions) {
+            if (in_array(PermissionProvider::PERMISSION_GRANT, $permissions, true)) {
+                return;
+            }
+        }
+
+        throw new PermissionsServiceException(
+            sprintf(
+                'Resource %s should have at least one user with GRANT access',
+                $resourceId
+            )
+        );
+    }
+
+    private function wetRun(array $resources): void
+    {
+        $this->dataBaseAccess->removeMultiplePermissionsNew($resources);
+        $this->dataBaseAccess->addMultiplePermissionsNew($resources);
+    }
+
+    private function triggerEvents(
+        core_kernel_classes_Resource $resource,
+        array $permissionsDelta,
+        bool $isRecursive
+    ): void {
+        $this->queueDispatcher->createTask(
+            new PostChangePermissionsTask(),
+            [
+                PostChangePermissionsTask::PARAM_RESOURCE_ID => $resource->getUri(),

```

Comment:
This will be always the root resource, correct? So it wont be misleading in the event that permission was only updated for this resource instead of the whole tree? Oh, I see we do it using the parameter `PARAM_IS_RECURSIVE` which was introduced in our last refactoring. Maybe this ambiguity could be removed too. Not sure we want to `is_recursive` to be controlled anymore as we do not want multiple tasks anymore :) 

Wdyt?

Reply:
Task was removed